clients/drcachesim/tests/trace_interval_analysis_unit_tests.cpp

```diff
@@ -0,0 +1,246 @@
+            if (record.marker.type == TRACE_TYPE_MARKER &&
+                record.marker.marker_type == TRACE_MARKER_TYPE_TIMESTAMP) {
+                last_timestamp_ = record.marker.marker_value;
+                if (first_timestamp_ == 0)
+                    first_timestamp_ = last_timestamp_;
+            }
+            return scheduler_t::stream_status_t::STATUS_OK;
+        }
+        return scheduler_t::stream_status_t::STATUS_EOF;
+    }
+    std::string
+    get_stream_name() const override
+    {
+        return "test_stream";
+    }
+    scheduler_t::input_ordinal_t
+    get_input_stream_ordinal() override
+    {
+        return 0;
+    }
+
+private:
+    std::vector<memref_t> refs_;
+    size_t at_;
+};
+
+// Test analyzer_t that uses a test_stream_t instead of a stream provided
+// by a scheduler_t.
+class test_analyzer_t : public analyzer_t {
+public:
+    test_analyzer_t(const std::vector<memref_t> &refs, analysis_tool_t **tools,
+                    int num_tools, bool parallel, uint64_t interval_microseconds)
+        : analyzer_t()
+    {
+        num_tools_ = num_tools;
+        tools_ = tools;
+        parallel_ = parallel;
+        interval_microseconds_ = interval_microseconds;
+        verbosity_ = 2;
+        worker_count_ = 1;
+        test_stream_ = std::unique_ptr<scheduler_t::stream_t>(new test_stream_t(refs));
+        worker_data_.push_back(analyzer_worker_data_t(0, test_stream_.get()));
+    }
+
+private:
+    std::unique_ptr<scheduler_t::stream_t> test_stream_;
+};
+
+// Test analysis_tool_t that stores information about when the interval-end
+// events (generate_shard_interval_result and generate_interval_result) were invoked.
+class test_analysis_tool_t : public analysis_tool_t {
+public:
+    test_analysis_tool_t()
+        : seen_memrefs_(0)
+        , seen_parallel_memrefs_(0)
+    {
+    }
+    bool
+    process_memref(const memref_t &memref) override
+    {
+        ++seen_memrefs_;
+        return true;
+    }
+    bool
+    generate_interval_result(uint64_t interval_id) override
+    {
+        serial_interval_ends_.push_back(std::make_pair(interval_id, seen_memrefs_));
+        return true;
+    }
+    bool
+    print_results() override
+    {
+        return true;
+    }
+    bool
+    parallel_shard_supported() override
+    {
+        return true;
+    }
+    void *
+    parallel_shard_init(int shard_index, void *worker_data) override
+    {
+        return reinterpret_cast<void *>(kShardData);
+    }
+    bool
+    parallel_shard_exit(void *shard_data) override
+    {
+        return true;
+    }
+    bool
+    parallel_shard_memref(void *shard_data, const memref_t &memref) override
+    {
+        ++seen_parallel_memrefs_;
+        return true;
+    }
+    bool
+    generate_shard_interval_result(void *shard_data, uint64_t interval_id) override
+    {
+        if (shard_data != reinterpret_cast<void *>(kShardData)) {
+            fprintf(stderr, "Invalid shard_data\n");
```

Comment:
FATAL_ERROR

Reply:
I had not used FATAL_ERROR on purpose, thinking that it's better for the error to be propagated up using the `false` returned value.

Reply:
But maybe it is better to just fail immediately. Done.